// symbol-sdk/dist/src/model/namespace/NamespaceInfo.js
"use strict";
/*
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamespaceInfo = void 0;
const catbuffer_typescript_1 = require("catbuffer-typescript");
/**
 * Object containing information of a namespace.
 */
class NamespaceInfo {
    /**
     * @param active
     * @param index
     * @param recordId
     * @param registrationType
     * @param depth
     * @param levels
     * @param parentId
     * @param ownerAddress
     * @param startHeight
     * @param endHeight
     */
    constructor(
    /**
     * Namespace is active.
     */
    active, 
    /**
     * The namespace index.
     */
    index, 
    /**
     * The meta data id.
     */
    recordId, 
    /**
     * The namespace registration type, namespace and sub namespace.
     */
    registrationType, 
    /**
     * The level of namespace.
     */
    depth, 
    /**
     * The namespace id levels.
     */
    levels, 
    /**
     * The namespace parent id.
     */
    parentId, 
    /**
     * The namespace owner's address.
     */
    ownerAddress, 
    /**
     * The height at which the ownership begins.
     */
    startHeight, 
    /**
     * The height at which the ownership ends.
     */
    endHeight, 
    /**
     * The alias linked to a namespace.
     */
    alias) {
        this.active = active;
        this.index = index;
        this.recordId = recordId;
        this.registrationType = registrationType;
        this.depth = depth;
        this.levels = levels;
        this.parentId = parentId;
        this.ownerAddress = ownerAddress;
        this.startHeight = startHeight;
        this.endHeight = endHeight;
        this.alias = alias;
    }
    /**
     * Namespace id
     * @returns {Id}
     */
    get id() {
        return this.levels[this.levels.length - 1];
    }
    /**
     * Is root namespace
     * @returns {boolean}
     */
    isRoot() {
        return this.registrationType === 0;
    }
    /**
     * Is sub namepsace
     * @returns {boolean}
     */
    isSubnamespace() {
        return this.registrationType === 1;
    }
    /**
     * Has alias
     * @returns {boolean}
     */
    hasAlias() {
        return this.alias.type !== 0;
    }
    /**
     * Get parent id
     * @returns {Id}
     */
    parentNamespaceId() {
        if (this.isRoot()) {
            throw new Error('Is a Root Namespace');
        }
        return this.parentId;
    }
    /**
     * Generate buffer
     * @return {Uint8Array}
     */
    serialize(children) {
        const id = this.id.toBuilder();
        const ownerAddress = this.ownerAddress.toBuilder();
        const lifetime = new catbuffer_typescript_1.NamespaceLifetimeBuilder(new catbuffer_typescript_1.HeightDto(this.startHeight.toDTO()), new catbuffer_typescript_1.HeightDto(this.endHeight.toDTO()));
        const rootAlias = this.alias.type;
        const paths = children.map((dto) => this.toNamespaceAliasTypeDto(dto));
        return new catbuffer_typescript_1.RootNamespaceHistoryBuilder(id, ownerAddress, lifetime, rootAlias, paths).serialize();
    }
    toNamespaceAliasTypeDto(namespaceInfo) {
        var _a, _b;
        const path = namespaceInfo.levels.map((id) => id.toBuilder());
        const alias = new catbuffer_typescript_1.NamespaceAliasBuilder(namespaceInfo.alias.type.valueOf(), (_a = namespaceInfo.alias.mosaicId) === null || _a === void 0 ? void 0 : _a.toBuilder(), (_b = namespaceInfo.alias.address) === null || _b === void 0 ? void 0 : _b.toBuilder());
        return new catbuffer_typescript_1.NamespacePathBuilder(path, alias);
    }
}
exports.NamespaceInfo = NamespaceInfo;
